Add travis integration

KeyserSosa opened the pull request:
The only way to test travis is to deploy to master with a `.travis.yml` file and wait for travis-ci to pick up the change.  To do some pre-testing here, the `Vagrantfile` has been split into two distinct machines:

 * `default` is, as before, the full reddit build (we have an opportunity to rename this)
 * `travis` is a minimal install with only the packages and services required to run `nosetests` against the codebase.  It tries to mimic what will happen when we try to build this on travis-ci's workers.  

As part of this addition, I've moved `install-reddit.sh` to `install/reddit.sh` and populated that `install` folder with common scripts used for both `default` and `travis` builds.

KeyserSosa commented:
:eyeglasses: @bsimpson63 @spladug @ketralnis 

Most of the new shell scripts are just copy-paste from the original `install-reddit.sh`.  For the remaining bits, I used all ten of my thumbs.  

The sticky points were either around permissions or removing stock packages that travis installs by default.  Specifically:

 * There is a script to downgrade cassandra, since their default is 2.0
 * There are some `dpkg` and `apt` acropatics (aptobatics?) to uninstall their libmemcached and substitute our own.  

Most importantly: 

![screenshot 2016-01-12 16 21 53](https://cloud.githubusercontent.com/assets/100474/12281279/a0a6fd62-b949-11e5-9884-ebcf4969b907.png)

jedberg commented:
Continuous integration for reddit?  Did I just see flying bacon? :P

bsimpson63 commented on Vagrantfile:
```diff
@@ -104,49 +104,66 @@ Vagrant.configure(2) do |config|
-      echo "additional setup already run"
-    fi
-  SCRIPT
+  config.vm.define "default" do |redditlocal|
```
note to self: will need to do `vagrant up default` to name the VM and ensure that this block runs

bsimpson63 commented on Vagrantfile:
```diff
@@ -104,49 +104,66 @@ Vagrant.configure(2) do |config|
-    config.vm.provision "shell",
-      path: "#{code_share_host_path}/private/vagrant_setup.sh",
-      args: [vagrant_user]
+  config.vm.define "travis" do |travis|
```
Can you add a comment here or at the top of the file explaining that this is never run, but is to try to replicate what will happen in travis?

bsimpson63 commented on install/setup_postgres.sh:
```diff
@@ -0,0 +1,53 @@
+PGSCRIPT
+fi
+
+sudo -u postgres psql reddit <  $FUNCTIONSQL
```
Don't feel strongly, but you could move all the stuff in functions.sql into this file to avoid needing to pass the path. It doesn't look like anything else is relying on functions.sql.

KeyserSosa replied:
This is really obvious in retrospect.  Fixing now!

bsimpson63 commented:
Maybe not worth worrying about yet, but how will plugins work with this?

bsimpson63 commented on .travis.yml:
```diff
@@ -0,0 +1,33 @@
+    system_site_packages: true
+
+services:
+    - cassandra
```
Is it just much faster to include cassandra as a service rather than installing it in travis.sh, even with the downgrade?

KeyserSosa replied:
Good point.  I'll give it a try without this line to see if it can still attach to a by-hand allocated cassandra service.  I can't see why it wouldn't.

bsimpson63 commented on install/travis_downgrade_cassandra.sh:
```diff
@@ -0,0 +1,37 @@
+CASSPACKAGE=apache-cassandra-$CASSANDRA_VERSION
+
+sudo rm -rf /var/lib/cassandra
+wget http://archive.apache.org/dist/cassandra/$CASSANDRA_VERSION/$CASSPACKAGE-bin.tar.gz
```
no way to use the datastax version?

bsimpson63 commented:
:nail_care: looks great, just a couple questions

KeyserSosa commented:
:haircut:

spladug commented:
To keep [cool URLs in docs etc.](https://www.w3.org/Provider/Style/URI.html) could `install-reddit.sh` still exist and just link to `install/reddit.sh`?

spladug commented on Vagrantfile:
```diff
@@ -25,7 +25,15 @@
+#              up the build.  This will *not* be the same environment as
+#              travis, but it should be useful for repairing broken tests.
+#
+# To start your vagrant box simply enter `vagrant up default` from {ROOTDIR}/reddit.
```
Does this mean people will have to add "default" every time they want to up a default install? Could that be... defaulted for ergonomics?

KeyserSosa replied:
On closer reading of the [Vagrant docs](https://docs.vagrantup.com/v2/multi-machine/) it looks like yes.  [Unfortunately they don't say as much until several sections after they say "Commands that only make sense to target a single machine, such as vagrant ssh, now *require* the name of the machine to control." which initially threw me.]

spladug commented on install/done.sh:
```diff
@@ -0,0 +1,88 @@
+set -e
+
+###############################################################################
+# Configuration
```
this doesn't really make sense as a configuration section anymore. these variable should be coming from elsewhere and the flow of just reading the script to find and modify them is gone.

spladug commented on install/install_services.sh:
```diff
@@ -0,0 +1,65 @@
+PACKAGES
+
+# we don't want to upgrade to C* 2.0 yet, so we'll put it on hold
+apt-mark hold cassandra || true
```
this should probably be in the cassandra script?

spladug commented on install/install_cassandra.sh:
```diff
@@ -0,0 +1,45 @@
+wget -qO- -L https://debian.datastax.com/debian/repo_key | sudo apt-key add -
+
+# grab the new ppas' package listings
+sudo apt-get update
```
it's kinda a bummer that we'll be calling `apt-get update` so many times now. that's quite slow.

KeyserSosa replied:
Attempted to refactor away one in 230b1ed.  It adds a dependency between two of the scripts, though.  Let me know if that seems reasonable.

spladug commented on install/install_apt.sh:
```diff
@@ -0,0 +1,104 @@
+set -x -e
+
+# aptitude configuration
+APTITUDE_OPTIONS=${APTITUDE_OPTIONS:-"-y"}
```
perhaps these scripts could source a common config file rather than duplicating config like this?

KeyserSosa replied:
Done.  I wasn't sure of preferred style for this.  e20cca6.

spladug commented:
:no_bell: pet peeve: "setup" is a noun not a verb!

spladug commented:
:nail_care: i like splitting up the script, but i think there's still some work to reduce duplication and maintain ergonomics. awesome work getting it to build on travis, loving those green checkmarks!

KeyserSosa commented:
:haircut: 

@spladug I think I addressed everything except 'setup' (though I at least fixed the comments that I added).  Do you want me to rename the scripts `set_up_*` or turn it into a suffix `_setup.sh`?

bsimpson63 commented on install/install_cassandra.sh:
```diff
@@ -0,0 +1,49 @@
+sudo service cassandra start
+
+# check each port for connectivity
+echo "Waiting for cassandra to be available, see source for port meanings..."
```
the part about "see source for port meanings" isn't necessary anymore

bsimpson63 commented on install/reddit.sh:
```diff
@@ -0,0 +1,639 @@
+$RUNDIR/setup_cassandra.sh
+
+# Configure PostgreSQL
+$RUNDIR/setup_postgres.sh $REDDIT_SRC/reddit
```
don't need the path anymore

bsimpson63 commented on install/install_cassandra.sh:
```diff
@@ -0,0 +1,49 @@
+fi
+
+# grab the new ppas' package listings
+sudo apt-get update
```
hasn't this already run?

bsimpson63 commented:
:fish: lgtm

spladug commented:
:fish: looks great! thanks for all that
